src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -316,8 +316,30 @@ private async Task PullContainerAsync(IExecutionContext executionContext, Contai
                 ArgUtil.NotNullOrEmpty(username, nameof(username));
                 ArgUtil.NotNullOrEmpty(password, nameof(password));
 
-                int loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
-                if (loginExitCode != 0)
+
+                int retryCount = 0;
+                int loginExitCode = 0;
+
+                while (retryCount < 3)
```

Comment:
would be nice to pull 3 up into an inline const (const int maxTries = 3) to prevent typos if that's ever changed